### ISSUE TITLE
Bulk actions should be consistent about opening versions (or not)

### ISSUE THREAD

**Issue by ndushay** [sul-dlss/argo#3683]

**Comment by justinlittman** [comment]
There have been significant changes in bulk action handling of versioning since this ticket was created. Are there any specific bulk actions that seem inconsistent?

**Comment by estanfor** [comment]
@justinlittman I've added more information in the ticket. @andrewjbtw could you review to see if you agree?

**Comment by justinlittman** [comment]
This is based on a code read:

| Action      | Opens version      | Closes version |
| ------------- | ------------- | ------------- |
| Add workflow | No | No |
| Text extraction | Workflow will open version if not open | Workflow will close version |
| Refresh metadata from FOLIO record | Yes | No |
| Upload descriptive metadata spreadsheet | Yes, if needed | No |
| Import FOLIO Instance HRIDs, barcodes, and serials metadata | Yes | No |
| Update rights | Must be open | No |
| Update license and rights statements | Yes, if needed | No |
| Manage embargo | Yes, if needed | No |
| Update governing APO | Yes, if needed | No |
| Apply APO defaults | Yes, if needed | No |
| Update source IDs | Yes, if needed | No |
| Update content type | Must be open | No |
| Import structural metadata | Must be open | No |
| Update collection | Yes, if needed | No |
| Create virtual object | DSA opens if needed | DSA closes |

**Comment by andrewjbtw** [comment]
The descriptive metadata bulk action definitely closes the versions. I'm not sure how it's done - I don't think it's via workflow - but we put that in place when creating the cocina spreadsheet while leaving Fedora. There's been a longstanding issue with people not realizing that they have to close a version in order to publish metadata changes and auto-closing has mostly resolved that.

I like the idea of giving people an option to close versions - whether that's in this issue or a separate one. There are a number of use cases for making multiple changes to a set before closing the version.

**Comment by estanfor** [comment]
I'll create a separate issue for closing versions, and I can double-check the bulk actions with open/close discrepancies between our lists.

**Comment by estanfor** [comment]
@justinlittman OK, I think the "change the following bulk actions to open a new version if needed" list would be:

- Add workflow
- Update rights
- Update content type
- Import structural metadata

**Comment by andrewjbtw** [comment]
"Add workflow" is an odd one out. I would like to eventually get rid of it altogether, as it's now down to just six workflows after showing all of them.

I'm not familiar with how the GIS workflows operate if invoked from Argo and we should check if they're still needed. For many years, those workflows were the only way to carry out certain actions, but since GIS got integrated with Preassembly they may no longer be necessary. I think they require not just a new version but some actions to be taken on a filesystem too.

The other workflows:

- goobiWF -- definitely requires an open version, should not close the version after running
    - the function of this WF is to notify goobi that a druid exists
    - it's usually run at registration time but has to be added later if a druid wasn't started with a Goobi deposit
    - version is closed later when Goobi accessions the next version
- wasDisseminationWF -- I am not sure there's any reason for this to be an option
- wasCrawlPreassemblyWF -- also not sure there's a reason for this, used to be necessary for hacking together manual WARC accessioning
- wasSeedPreassemblyWF -- still need this one
  - this is the only way to "regenerate" a seed thumbnail
  - it will automatically run accessionWF and close the version

**Comment by estanfor** [comment]
@justinlittman What's the distinction between "Yes" and "Yes, if needed"? The two plain "Yes"es in your list (Refresh from FOLIO and Import FOLIO...) do require an open version in order to run, as far as I know (as does everything that's marked "Yes, if needed", I *think*)

@andrewjbtw would you prefer to leave "Add workflow" out of this issue?

**Comment by justinlittman** [comment]
"Yes, if needed": Check if a version is already open. If not, open it.
"Yes": Try to open a version. (Presumably, it is an error if already open.)

**Comment by estanfor** [comment]
OK, then I would say all of them should be "Yes, if needed".

**Comment by andrewjbtw** [comment]
Agreed - I can't think of a situation where it's a problem if an object someone wants to open is already opened.

**Comment by estanfor** [comment]
I'd like to have it on the "Yes, if needed" list just so that we can say that you never need to separately open versions before running a bulk action, but I'll defer to @andrewjbtw's expertise.
